src/funnels/Swapper.sol

```diff
@@ -71,21 +72,34 @@ contract Swapper {
         emit Deny(usr);
     }
 
-    function setLimits(address src, address dst, uint64 hop, uint128 cap) external auth {
+    function setLimits(address src, address dst, uint128 cap, uint64 hop) external auth {
+        uint128 amt = limits[src][dst].amt;
         limits[src][dst] = PairLimit({
-            hop:  hop,
-            zzz:  limits[src][dst].zzz,
-            cap: cap
+            cap: cap,
+            hop: hop,
+            amt: amt > cap ? cap : amt,
+            zzz: limits[src][dst].zzz
         });
-        emit SetLimits(src, dst, hop, cap);
+        emit SetLimits(src, dst, cap, hop);
     }
 
     function swap(address src, address dst, uint256 amt, uint256 minOut, address callee, bytes calldata data) external auth returns (uint256 out) {
         PairLimit memory limit = limits[src][dst];
-        require(block.timestamp >= limit.zzz + limit.hop, "Swapper/too-soon");
-        limits[src][dst].zzz = uint64(block.timestamp);
 
-        require(amt <= limit.cap, "Swapper/exceeds-max-amt");
+        unchecked {
+            if (block.timestamp - limit.zzz >= limit.hop) {
+                // Reset batch
+                limit.amt = limit.cap;
+                limit.zzz = uint64(block.timestamp);
+            }
+        }
+
+        require(amt <= limit.amt, "Swapper/exceeds-max-amt");
+
+        unchecked {
+            limits[src][dst].amt = limit.amt - uint128(amt);
+            limits[src][dst].zzz = limit.zzz;
```

Comment:
I know we said these 2 lines are grouped to one storage op, but for some reason I see that when commenting out the 2nd line we get a 20K gas saving (180359 for just the first swap in testSwap).

Also note that the base code takes 181131 gas for the first swap in testSwap, while the altered one takes 206298.
I wouldn't expect such a big diff as I don't think there should be more storage write ops.

Reply:
I think this is understood now. Commenting the `zzz` makes a change because in this specific test is the only variable actually changing. So at least the first part should be solved now, not sure if I'm following the second paragraph.

Reply:
For the second one, after discussing I realized it has to be due to a storage warming happening on `setLimits` called by the `setUp` in the old structure, as the only slot gets written. However in the new structure the slot having the `amt/zzz` shouldn't be written in `setLimits` as `amt <= cap`.